### PR TITLE
Terraform and load script

### PR DESCRIPTION
**Pull request by caithomas6789**
- ECS service for dashboard is not working but is close.
- Reworking the entire load script.

### CODE REVIEW

File: pipeline/load.py
Comment (IB64) [inline pipeline/load.py:1]: Consistent type hinting through out!

File: pipeline/load.py [hunk 1]
```diff
@@ -21,130 +31,194 @@ def get_db_connection() -> extensions.connection:
...
-            if Levenshtein.normalized_similarity(
-                    genre[1], new_genre.lower()) > 0.75:
-                return genre[1]
-
-        new_genre = new_genre.replace("'", "''")
-        cur.execute(
-            f"INSERT INTO genre(genre) VALUES ('{new_genre.lower()}');")
-        db_connection.commit()
-        return new_genre
+        return {r[1]: r[0] for r in genres}
```
Comment (IB64) [inline pipeline/load.py:41]: Clever return

File: pipeline/load.py [hunk 2]
```diff
@@ -21,130 +31,194 @@ def get_db_connection() -> extensions.connection:
...
         cur.execute("SELECT * FROM country;")
 
         countries = cur.fetchall()
-        for country in countries:
-            if country[1] == new_country:
-                return
+        return {r[1]: r[0] for r in countries}
+
+
+def get_items(db_connection: extensions.connection) -> dict:
```
Comment (IB64) [inline pipeline/load.py:66]: Nice simple functions

File: pipeline/load.py [hunk 3]
```diff
@@ -21,130 +31,194 @@ def get_db_connection() -> extensions.connection:
...
+    Returns a dictionary of all the current items in the database.
+    """
+    with db_connection.cursor() as cur:
+        cur.execute("SELECT * FROM item;")
+
+        items = cur.fetchall()
+        return {r[2]: r[0] for r in items}
+
+
+def check_if_genre_in_db(new_genre: str, genres: dict) -> None:
```
Comment (caitlinturnidge) [inline pipeline/load.py:77]: nice broken down functions

File: pipeline/load.py [hunk 4]
```diff
@@ -155,20 +229,31 @@ def load_sales_event(new_sale: pd.Series, db_connection: extensions.connection)
...
-    sale_df_flat_tags.apply(load_items, db_connection=con, axis=1)
-    print("Items added")
-    sale_df_flat_tags.apply(load_item_genres, db_connection=con, axis=1)
-    print("Item genres added")
-    sale_df.apply(load_sales_event, db_connection=con, axis=1)
-    print("Sales added")
+    db_genres = get_genres(db_connection)
+    db_artists = get_artists(db_connection)
+    db_countries = get_countries(db_connection)
+    db_items = get_items(db_connection)
```
Comment (IB64) [inline pipeline/load.py:239]: Clever to call it all at once!

File: pipeline/load.py [hunk 5]
```diff
@@ -21,130 +31,194 @@ def get_db_connection() -> extensions.connection:
...
+        query = f"""
+            INSERT INTO artist(artist_name) VALUES (%s) ON CONFLICT DO NOTHING;
+            """
 
-        for item in items:
-            if item[2] == new_item['title']:
-                return
+        cur.executemany(query, list)
+        db_connection.commit()
+        print("Artists added!")
```
Comment (caitlinturnidge) [inline pipeline/load.py:150]: Could add logging in the future? then we can see times as well

**Review by IB64 — APPROVED** [review]
Nice

File: pipeline/load.py [hunk 6]
```diff
@@ -155,20 +229,31 @@ def load_sales_event(new_sale: pd.Series, db_connection: extensions.connection)
...
-    print("Sales added")
+    db_genres = get_genres(db_connection)
+    db_artists = get_artists(db_connection)
+    db_countries = get_countries(db_connection)
+    db_items = get_items(db_connection)
+
+    flat_dataframe['tags'].apply(check_if_genre_in_db, genres=db_genres)
+    add_genres_to_database(db_connection, GENRES_NOT_IN_DB)
+
+    flat_dataframe['artist'].apply(check_if_artist_in_db, artists=db_artists)
```
Comment (caitlinturnidge) [inline pipeline/load.py:244]: could we add parallel apply here? might be quicker?

File: pipeline/pipeline.py [hunk 7]
```diff
@@ -6,24 +6,26 @@
...
 
     # Load
     load_dotenv()
     con = get_db_connection()
-    load_data(clean_data_exploded, clean_data, con)
-    print("Loaded!")
+    load(con, clean_data_exploded, clean_data)
+    print("Loaded!", perf_counter() - start)
 
     print(f"Time taken: {perf_counter() - start}")
```
Comment (caitlinturnidge) [inline pipeline/pipeline.py:31]: Good job on making it quicker!

**Review by caitlinturnidge — APPROVED** [review]

**Pull request merged**